dbms/src/DataStreams/CastTypeBlockInputStream.cpp

```diff
@@ -29,98 +26,91 @@ String CastTypeBlockInputStream::getName() const
-
-            FunctionPtr & cast_function = cast_functions[i];
-
-            /// Initialize function.
-            if (!cast_function)
-            {
-                cast_function = FunctionFactory::instance().get("CAST", context);
-
-                DataTypePtr unused_return_type;
-                ColumnsWithTypeAndName arguments{ temporary_block.getByPosition(0), temporary_block.getByPosition(1) };
-                std::vector<ExpressionAction> unused_prerequisites;
-
-                /// Prepares function to execution. TODO It is not obvious.
-                cast_function->getReturnTypeAndPrerequisites(arguments, unused_return_type, unused_prerequisites);
-            }
-
-            cast_function->execute(temporary_block, {0, 1}, 2);
-
-            res.insert({
-                temporary_block.getByPosition(2).column,
-                cast_types[i]->type,
-                cast_types[i]->name});
+            res.insert(src_column);
         }
         else
         {
-            res.insert(elem);
+            CastElement & cast_element = it->second;
+
+            size_t tmp_col = cast_element.tmp_col_offset;
+            ColumnNumbers arguments{tmp_col, tmp_col + 1};
+            tmp_conversion_block.getByPosition(tmp_col).column = src_column.column;
+
+            cast_element.function->execute(tmp_conversion_block, arguments, tmp_col + 2);
+            res.insert(tmp_conversion_block.getByPosition(tmp_col + 2));
         }
     }
 
     return res;
 }
 
-void CastTypeBlockInputStream::collectDifferent(const Block & in_sample, const Block & out_sample)
+
+CastTypeBlockInputStream::CastElement::CastElement(std::shared_ptr<IFunction> && function_, size_t tmp_col_offset_)
+    : function(std::move(function_)), tmp_col_offset(tmp_col_offset_) {}
+
+
+void CastTypeBlockInputStream::initialize(const Block & src_block)
 {
-    size_t in_size = in_sample.columns();
-    cast_types.resize(in_size);
-    for (size_t i = 0; i < in_size; ++i)
+    for (size_t src_col = 0; src_col < src_block.columns(); ++src_col)
     {
-        const auto & in_elem  = in_sample.getByPosition(i);
-        const auto & out_elem = out_sample.getByPosition(i);
+        const auto & src_column = src_block.getByPosition(src_col);
+
+        /// Skip, if it is a problem, it will be detected on the next pipeline stage
+        if (!ref_defenition.has(src_column.name))
+            continue;
+
+        const auto & ref_column = ref_defenition.getByName(src_column.name);
 
         /// Force conversion if source and destination types is different.
-        if (!out_elem.type->equals(*in_elem.type))
+        if (!ref_column.type->equals(*src_column.type))
         {
-            cast_types[i] = NameAndTypePair(out_elem.name, out_elem.type);
+            ColumnWithTypeAndName src_columnn_copy = src_column.cloneEmpty();
+            ColumnWithTypeAndName alias_column(std::make_shared<ColumnConstString>(1, ref_column.type->getName()), std::make_shared<DataTypeString>(), "");
```

Comment:
Difficult to read. This is not really an alias - it is const column with required type name.